packages/workbox-webpack-plugin/src/index.js

```diff
@@ -0,0 +1,181 @@
+  /**
+   * Creates an instance of WorkboxWebpackPlugin.
+   *
+   * @param {module:workbox-build.Configuration} [config] All the options as
+   *        passed to {@link module:workbox-build.generateSWString}.
+   * @param {Array<String>} [config.chunks] Array of chunk names to use for
+   *        generating the asset manifest. All assets belonging to the provided
+   *        chunk names will be included in the asset manifest. Any chunks that
+   *        are not listed or do not have a name will be removed.
+   * @param {Array<String>} [config.excludeChunks] Array of chunk names to
+   *        exclude from the asset manifest. Any asset beloning to the provided
+   *        chunk names will not be included in the asset manifest. This does
+   *        not affect chunks with no chunk name.
+   * @param {string} [config.filename = 'sw.js'] Name of the service worker file
+   * @param {string} [config.manifestFilename = 'file-manifest[hash].js'] Name
+   *        of the manifest file that will be written to the build directory
+   * @param {string} [config.swSrc] Path to an existing service worker file.
+   *        Will be added to the webpack compilation and prepended with
+   *        importScripts('workbox-sw.js', 'file-manifest.js')
+   */
+  constructor(config = {}) {
+    this.config = config;
+  }
+
+  /**
+   * @private
+   * @return {Object} All workbox configuration options that can be accepted
+   * by {@link module:workbox-build.generateSWString}
+   */
+  get generateSWStringOptions() {
+    const {
+      importScripts = [],
+      // TODO: decide on other options that should be passed to
+      // `generateSWString`. What operations should be possible for webpack
+      // users? eg, should webpack users also be able to specify a
+      // `globDirectory` that allows manifestEntries to be added that are
+      // outside webpack's scope?
+    } = this.config;
+
+    return {
+      importScripts,
+    };
+  }
+
+  /**
+   * @private
+   * @param {Object} [compiler] default compiler object passed from webpack
+   */
+  apply(compiler) {
+    /**
+     * The plugin was instanciated and the webpack compilation has just begun
+     * we configure the workbox-webpack-plugin/utils/read-file module to use
+     * webpack's compilation.inputFileSystem._readFile method for reading files
+     *
+     * TODO: Determine if this is absolutely necessary. It might be possible to
+     * only do this in development (when the file system is a "memory" file
+     * system). If that is the case, it might be better to set different values
+     * for setReadFile using compiler.plugin('run') for production and
+     * compiler.plugin('watch-run') for development.
+     */
+    setReadFile(compiler.inputFileSystem._readFile);
+    /**
+     * During the make phase of the webpack compilation, we use
+     * workbox-webpack-plugin/utils/copy-workbox-sw to add a built version of
+     * workbox-sw to the webpack compilation assets array.
+     */
```

Comment:
Hmm, I think this is not what everyone would want. 
We build `service-worker` separately by another webpack build, using workbox from npm which is embed into service-worker file. We have bunch of custom loaders that modifies/optimizes/analyzes code, so we would not prefer to have it this way. This approach is also uncommon in webpack/js community.

Reply:
Specifying the path to the service worker you create in `swSrc` would use that service worker then prepend it with `importScripts('workbox-sw.prod.v2.x.x.js', 'file-manifest.js')` so then you can just run:

`workboxBuild.precache(__precacheManifest)` in your service worker.

Not sure if this is the best flow...

see https://github.com/GoogleChrome/workbox/pull/808/files/a2dcaf4235f9f10456988a78759eb17dfedf6823#diff-5b99f4463e518de9c1c1cc612e8f6b59R73 for the swSrc parameter

Reply:
I assumed a raw `service-worker.js` would look as follows:

```
import WorkboxSW from 'workbox-sw';
const workboxSW = new WorkboxSW();
workboxSW.precache(self.__precacheManifest);
```

After building with webpack and processing with workbox:

```
importScripts('file-manifest.{hash}.js');
(function () {
      /* webpack generated code here */
});
```

Ideally, users can opt to use `workbox-precaching` package instead of the whole `workbox-sw` when they don't need anything else than simple precaching.

Reply:
FWIW, we're expecting it to be common for developers to refer to `workbox-sw` from our new CDN URL, support for which is baked in to `workbox-build` v3. The CDN `workbox-sw` basically acts as a lightweight loader + API surface, and will use `importScripts` at runtime to pull in additional dependencies for things like routing, caching, etc.

That being said, we know that not every developer will want to rely on the CDN copy, and Webpack users in particular are in a good position to bundle an optimized copy of `workbox-sw` + the additional routing, caching, etc. code and serve it locally.

CC: @philipwalton who's also exploring the ergonomics of bundling Workbox code.

Reply:
This will need some more consideration. I think we need to decide if `workbox-sw.js` should be copied into the webpack assets array, use the CDN url or expect the developer to import the `workbox-sw` file into a `service-worker` webpack entry point. Maybe provide an option in the config to specify what option we should use and select a smart default... not sure what *smart* would be ;)